### PR TITLE
 C# Strings - Format Strings Block

### PR DESCRIPTION
**Pull request by hDoubleE**
Supported by dotnet/samples[#462](https://github.com/dotnet/samples/pull/462)
Modified the Format Strings section of [C# Strings](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/strings/) using @pkulikov suggestions from #9002 . This is a learning curve for me so I encourage feedback and modifications to writing as needed. 

Samples: Should these be in the same PR or a different one? I have a sample file for interpolation I can include if needed. Also, should the code snips for the two methods output the same for contrast or just keep System.Format sample as is. I didn't want to proceed with the samples without clarification. 

#9002 (a start)

### CODE REVIEW

File: docs/csharp/programming-guide/strings/index.md
Comment (BillWagner) [inline docs/csharp/programming-guide/strings/index.md]: Let's make this a relative link to the markdown.  Instead of "https://docs.microsoft.com..../interpolated-strings", use "../../tutorials/intro-to-csharp/interpolated-strings.md"

File: docs/csharp/programming-guide/strings/index.md
Comment (BillWagner) [inline docs/csharp/programming-guide/strings/index.md]: This is good. Consider changing "but represents an improvement due to ease of use and inline clarity" to "but improves ease of use and inline clarity".

File: docs/csharp/programming-guide/strings/index.md
Comment (BillWagner) [inline docs/csharp/programming-guide/strings/index.md]: Instead of "[format method](https://...)" it would be better to use the xref syntax:  `<xref:System.String.Format%2A?displayProperty=nameWithType>`

File: docs/csharp/programming-guide/strings/index.md
Comment (pkulikov) [inline docs/csharp/programming-guide/strings/index.md]: typo: "...to create interpolated strings" -> "...to create format strings"
And, maybe, "used" can be dropped.
Also a style nit: a sentence should finish with a period, not a colon. A possible fix would sound:
>There are two methods to create format strings: string interpolation and composite formatting.

(the `string.Format` method is the example of the composite formatting feature).

Then, H3 sections give more information about listed methods.

File: docs/csharp/programming-guide/strings/index.md
Comment (pkulikov) [inline docs/csharp/programming-guide/strings/index.md]: The first relative link must include the `.md` file extension (you also can make it shorter by reaching the "csharp" folder and then descending):
`[*interpolated strings*](../../language-reference/tokens/interpolated.md)`

File: docs/csharp/programming-guide/strings/index.md
Comment (pkulikov) [inline docs/csharp/programming-guide/strings/index.md]: "...improves the syntax of C# code." sounds strange. Also try to avoid gerunds ("utilizing"). Consider:
> Use string interpolation to improve readability and maintainability of your code.

**Comment by hDoubleE** [comment]
Thanks for the help and feedback @BillWagner and @pkulikov. I'm going to work through the revisions you've suggested and create a PR for the samples. I'm moonlighting this and am a first timer, so bear with me. I thought if I'm going to try and learn the process it might as well be on a change with some substance. Thanks for the support and the style tips. I'll do my best to keep the work moving.

File: docs/csharp/programming-guide/strings/index.md
Comment (hDoubleE) [reply docs/csharp/programming-guide/strings/index.md]: I noted your xref has a displayProperty and all the others on the page end at 2A. Just curious what specific effect this will have?

**Comment by hDoubleE** [comment]
I went through and made each suggested change. Thanks for all the great input. Please review for accuracy while I work up the samples. Is it preferable to recycle samples from other pages or generate new material? I'm leaning toward making my own, but maybe it's better to stick with whats available?

File: docs/csharp/programming-guide/strings/index.md
Comment (BillWagner) [reply docs/csharp/programming-guide/strings/index.md]: the "displayProperty=nameWithType" means the type and the member are displayed. 

All the keys should be in the [Contributor Guide](https://docs.microsoft.com/contribute/dotnet-style-guide#links-to-apis)

File: docs/csharp/programming-guide/strings/index.md
Comment (pkulikov) [inline docs/csharp/programming-guide/strings/index.md]: Both relative links are broken: they suffer from one excessive `../`. Should be: `(../../language-reference...)` and `(../../tutorials...)`

**Comment by hDoubleE** [comment]
Hopefully that fixes the relative link paths. I also changed an H3 from "Format Method" to "Composite Formatting" after talking to @pkulikov, it seemed more technically accurate. Agree? I'm working on getting the snippets PR up so we can close this one out. Thanks for all the help!

File: docs/csharp/programming-guide/strings/index.md
Comment (pkulikov) [inline docs/csharp/programming-guide/strings/index.md]: @hDoubleE sorry, I've missed it earlier: the second relative link is still broken. This time, because of not correct file extension: it must be `.yml` instead of `.md`.

If you wonder how I see what is broken, here is the "magic":
image

Note the **OpenPublishing.Build** entry in the list of PR checks. If it has *Warnings* in status, select the [Details](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C11%5C16%5Ca3bda507-3390-de91-8f7f-26f90f4e5fc8%5CPullRequest%5C201811160105396336-9025%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=ZZMX1Pf0BbAzXa0C7whg5UHE0UuVV2ADMArxY%2BCzpXc%3D&st=2018-11-16T01%3A05%3A02Z&se=2018-12-17T01%3A10%3A02Z&sp=r) link to see the report. From that report, you can learn what goes not perfect during the build of the files from your PR (ignore all other files, though).

Comment (hDoubleE) [reply docs/csharp/programming-guide/strings/index.md]: Added .yml and now it links to file in VS code. Hopefully put that to rest. I reviewed all the great links. The index.md file has a local snippet folder in the docs repo. Although tempted, I did not add mine to it. I added to docs/samples repo and linked the path to that location. Please verify code snippet path for correctness. Hopefully the `<snippet>` and .md link correctly. Thank you.

**Comment by hDoubleE** [comment]
Still getting errors in Build: Cannot resolve '~/samples/snippets/csharp/programming-guide/strings/Strings_1.cs' relative to 'docs/csharp/programming-guide/strings/index.md'.

Please advise what action to take to clear this up. The 'strings' folder I made where Strings_1.cs resides in my local fork does not exist in dotnet/samples repo file structure.

File: docs/csharp/programming-guide/strings/index.md
Comment (hDoubleE) [reply docs/csharp/programming-guide/strings/index.md]: Thanks for the tip. I'll use this after pushing to verify.

**Comment by pkulikov** [comment]
> Please advise what action to take to clear this up.

Wait. This PR is built against the master branch of the samples repo. So, we need to wait until the supporting dotnet/samples#462 PR is merged. After it's merged, this PR would be closed/reopened to trigger a new build to check if new snippets are picked up.

Usually, after addressing all the feedback, contributors @-mention the reviewer in the comment to let them know that changes were addressed and PR is ready for the next round of review. That almost always makes the process faster.

File: docs/csharp/programming-guide/strings/index.md
Comment (pkulikov) [inline docs/csharp/programming-guide/strings/index.md]: @hDoubleE could you please delete this code file as well. As the code snippets are moved to the samples repo, there is no reason to keep that file in this repo.

File: docs/csharp/programming-guide/strings/index.md
Comment (pkulikov) [reply docs/csharp/programming-guide/strings/index.md]: > Although tempted, I did not add mine to it.

Great 👍

File: docs/csharp/programming-guide/strings/index.md
Comment (hDoubleE) [reply docs/csharp/programming-guide/strings/index.md]: So I accidentally deleted index.md and not index_6.cs. Really bad. Trying to figure out how to restore...

File: docs/csharp/programming-guide/strings/index.md
Comment (pkulikov) [inline docs/csharp/programming-guide/strings/index.md]: nit: "C# code" -> "your code" because of two reasons:
- C#-ness of the code is implicit because we are in the C# programming guide
- string interpolation is also available in Visual Basic, while current phrasing can be understood as if it's only C# feature.

File: docs/csharp/programming-guide/strings/index.md
Comment (pkulikov) [inline docs/csharp/programming-guide/strings/index.md]: "formatting C# code" -> "formatting .NET types"
Add the comma before "see".

**Comment by hDoubleE** [comment]
Saw change to review status, then back. What else needs to be done? Any way to include this change?

**Comment by BillWagner** [comment]
Thanks for the ping @hDoubleE 

I'll review it today, and if all's well, I'll :shipit:

**Comment by BillWagner** [comment]
closing and re-opening to force a build (and check the snippet includes)

**Review by BillWagner — APPROVED** [review]
Thanks for all the work, and all the updates on this @hDoubleE It's an impressive amount of work for your first PR with us. We are grateful to have community members like you working with us.  🌠 

Also, thanks to @pkulikov who answered questions and provided guidance.

I've reviewed all the changes, and I'll :shipit: now.

Thanks again for contributing to docs.

**Comment by hDoubleE** [comment]
Removed branch:

Correction: first PR ever!!! The learning curve was steep. But nothing like diving in to get started! Thanks @pkulikov for all your help and support and feedback throughout this process. I appreciate the chance to contribute!

One final thing for the group: What can I do better? What mistakes here (many) should I look to avoid in the future?

Thanks again. You guys rock.

**Pull request merged**
